### PR TITLE
Implementing VPG from scratch

### PR DESCRIPTION
**Pull request by Akashnil**
I'm planning to modify this to PPO / hyperparameter training later. For now, I'm changing the vanilla VPG algorithm to compute everything about the loss including advantages inside the computation graph, except for the trajectories themselves.

I tested it on 3 environments
CartPole, Swimmer, and FrozenLake. Initially it achieves performace similar to the vanilla VPG implementation, however after improving for 50-100 epochs, it suddenly drops down to meaningless policy and never recovers. I'm trying to investigate the issue.

### CODE REVIEW

File: spinup/algos/vpg/core.py [hunk 1]
```diff
@@ -27,6 +27,11 @@ def placeholders_from_spaces(*args):
     return [placeholder_from_space(space) for space in args]
 
 def mlp(x, hidden_sizes=(32,), activation=tf.tanh, output_activation=None):
+    # support for scalar inputs, eg. discrete observation spaces
+    if len(x.shape) == 1:
+        x = x[:,None]
+    if x.dtype not in [tf.bfloat16, tf.float16, tf.float32, tf.float64, tf.complex64, tf.complex128]:
+        x = tf.dtypes.cast(x, tf.float32)
```
[the remark this answers is not in the document]
Comment (Akashnil) [reply spinup/algos/vpg/core.py]: Done. Now there's one_hot encoding for discrete observations.

File: spinup/algos/vpg2/core.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    endmask = tf.reverse(endmask, axis=[axis])
+    p = list(range(len(x.get_shape())))
+    p[axis] = 0
+    p[0] = axis
+    x = tf.transpose(x, perm=p)
+    if last_val == None:
+        initializer = x[0] - x[0]
+    elif force_last:
+        # we want to ensure initializer * discount + x[0] == last_val
+        initializer = (last_val - x[0]) / discount
```
[the remark this answers is not in the document]
Comment (Akashnil) [reply spinup/algos/vpg2/core.py]: Later the initializer will get ignored if end-mask is 1 at the last value, in the line
`return cumsum * (1 - mask) * discount + x`
 That is intended.

File: spinup/algos/vpg2/vpg2.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+msk_buf_ph = tf.placeholder(dtype=tf.float32, shape=(None,))
+
+gamma = tf.constant(0.99)
+lam = tf.constant(0.95)
+
+pi_buf, logp_buf, logp_pi_buf, v_buf = actor_critic(obs_buf_ph, act_buf_ph, **ac_kwargs)
+ret_buf = core.masked_suffix_sum(rew_buf_ph, msk_buf_ph, gamma, axis=0, last_val=v_buf[-1])
+ret_buf = tf.stop_gradient(ret_buf)
+
+v_loss = tf.math.sqrt(tf.reduce_mean((ret_buf - v_buf)**2))
```
[the remark this answers is not in the document]
Comment (Akashnil) [reply spinup/algos/vpg2/vpg2.py]: So I wanted to have the unit of the loss be the same for value_loss and pi_loss, eg. if rewards are doubled, then both value_loss and pi_loss should get doubled. If the sqrt isn't applied then value_loss will be 4 times the original. I tried both sqrt and not sqrt, it didn't make much difference though.

File: spinup/algos/vpg2/vpg2.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+delta_buf = rew_buf_ph[:-1] + gamma * v_buf[1:] * (1 - msk_buf_ph[:-1]) - v_buf[:-1]
```
[the remark this answers is not in the document]
Comment (Akashnil) [reply spinup/algos/vpg2/vpg2.py]: current advantage = current_reward + discounted_future_rewards - current_value.

If current mask is 1, then discounted future rewards should go away. So mask is aligned with current_reward, not discounted_future_rewards.

**Comment by christopherhesse** [comment]
@Akashnil should this be closed?
